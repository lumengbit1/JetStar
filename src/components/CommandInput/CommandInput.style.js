import styled from 'styled-components';

export const Root = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 3.125rem;
  width: 100%;
`;

export const Input = styled.input`
  width: 30%;
  height: 1.875rem;
  font-size: 1.5rem;
`;