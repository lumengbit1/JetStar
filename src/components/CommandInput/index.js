import React from 'react';
import { Input } from './CommandInput.style';

const CommandInput = ()=>(
  <Input />
)

export default CommandInput;