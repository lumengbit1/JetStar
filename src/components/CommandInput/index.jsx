import React, { useCallback } from 'react';
import { useDispatch } from 'react-redux';
import { useImmer } from 'use-immer';
import { reset, handleCommand, clearErrorMessage } from '../../reducers/actions';
import { Root, Input, ButtonContainer, Button } from './CommandInput.style';

const CommandInput = () => {
  const [command, updateCommand] = useImmer('');
  const dispatch = useDispatch();

  const onChange = useCallback((e) => {
    e.preventDefault();

    updateCommand(e.target.value.toUpperCase());
  }, []);

  const handleSubmit = useCallback(
    (e) => {
      e.preventDefault();

      if (command.length === 0) return;

      dispatch(clearErrorMessage());

      dispatch(handleCommand(command));

      updateCommand('');
    },
    [command],
  );

  const handleReset = useCallback(
    (e) => {
      e.preventDefault();

      dispatch(reset());

      updateCommand('');
    },
    [command],
  );

  return (
    <Root>
      <Input
        value={command}
        aria-label="command-input"
        onChange={onChange}
        placeholder="Tell the robot what to do ..."
      />
      <ButtonContainer>
        <Button onClick={handleSubmit}>
          Run
        </Button>

        <Button onClick={handleReset}>
          Reset
        </Button>
      </ButtonContainer>
    </Root>
  );
};

export default CommandInput;
