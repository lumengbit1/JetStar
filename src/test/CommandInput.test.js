import React from 'react';
import 'jest-styled-components';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';

import { render, fireEvent } from '@testing-library/react';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store'
import CommandInput from '../components/CommandInput';

const mockStore = configureMockStore([thunk]);
const store = mockStore();

describe('CommandInput Test Cases', () => {
  it('1: Component shoule be rendered correctly', async () => {
    const { container } = render(
      <Provider store={store}>
        <CommandInput />
      </Provider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('2: Input shoule have correct value when type in', async () => {
    const { getByLabelText } = render(
      <Provider store={store}>
        <CommandInput />
      </Provider>,
    );

    const input = getByLabelText('command-input');

    fireEvent.change(input, {target: {value: 'MOVE'}})
    expect(input.value).toBe('MOVE')
  });
});