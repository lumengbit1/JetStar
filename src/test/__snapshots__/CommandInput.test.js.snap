// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommandInput Test Cases 1: Component shoule be rendered correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 3.125rem;
  width: 100%;
}

.c1 {
  width: 30%;
  height: 1.875rem;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin-top: 1rem;
}

.c3 {
  width: 5rem;
  height: 1.875rem;
}

.c3:first-child {
  margin-right: 1rem;
}

<div>
  <div
    class="c0"
  >
    <input
      aria-label="command-input"
      class="c1"
      placeholder="Tell the robot what to do ..."
      value=""
    />
    <div
      class="c2"
    >
      <button
        class="c3"
      >
        Run
      </button>
      <button
        class="c3"
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;
